worker/externalcontrollerupdater/externalcontrollerupdater.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package externalcontrollerupdater
+
+import (
+	"io"
+	"reflect"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/set"
+	"gopkg.in/juju/names.v2"
+	"gopkg.in/juju/worker.v1"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/crosscontroller"
+	"github.com/juju/juju/core/crossmodel"
+	"github.com/juju/juju/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+var logger = loggo.GetLogger("juju.worker.externalcontrollerupdater")
+
+type ExternalControllerUpdaterClient interface {
+	WatchExternalControllers() (watcher.StringsWatcher, error)
+	ExternalControllerInfo(controllerUUID string) (*crossmodel.ControllerInfo, error)
+	SetExternalControllerInfo(crossmodel.ControllerInfo) error
+}
+
+type ExternalControllerWatcherClientCloser interface {
+	ExternalControllerWatcherClient
+	io.Closer
+}
+
+type ExternalControllerWatcherClient interface {
+	WatchControllerInfo() (watcher.NotifyWatcher, error)
+	ControllerInfo() (*crosscontroller.ControllerInfo, error)
+}
+
+type NewExternalControllerWatcherClientFunc func(*api.Info) (ExternalControllerWatcherClientCloser, error)
+
+func New(
+	externalControllers ExternalControllerUpdaterClient,
+	newExternalControllerWatcherClient NewExternalControllerWatcherClientFunc,
+) (worker.Worker, error) {
+	watcher, err := externalControllers.WatchExternalControllers()
+	if err != nil {
+		return nil, errors.Annotate(err, "watching external controllers")
+	}
+
+	w := updaterWorker{
+		changes:                            watcher.Changes(),
+		externalControllerInfo:             externalControllers.ExternalControllerInfo,
+		setExternalControllerInfo:          externalControllers.SetExternalControllerInfo,
+		newExternalControllerWatcherClient: newExternalControllerWatcherClient,
+		runner: worker.NewRunner(worker.RunnerParams{
```

Comment:
I think we also need to add the runner to the catacomb of the main worker so it and all the subworkers are stopped when the main worker dies?

Reply:
It is, see the Init in catacomb.Plan below.

Reply:
Ah yes, so it is.